# Summary

Objective:

- Classify sections of land into 9 categories based on descriptions



# Part 1: Data Processing

a) import train and test data

In [2]:
import pandas as pd

In [3]:
import numpy as np

In [4]:
train_data = pd.read_csv(r'C:\Users\mhime\Downloads\train_data.csv')

In [5]:
train_data.head()

class  BrdIndx  Area  Round  Bright  Compact  ShpIndx  Mean_G  Mean_R  \
0  concrete      1.32   131   0.81  222.74     1.66     2.18  192.94  235.11   
1    shadow      1.59   864   0.94   47.56     1.41     1.87   36.82   48.78   
2    shadow      1.41   409   1.00   51.38     1.37     1.53   41.72   51.96   
3      tree      2.58   187   1.91   70.08     3.41     3.11   93.13   55.20   
4   asphalt      2.60   116   2.05   89.57     3.06     3.02   73.17   94.89   

   Mean_NIR  ...  SD_NIR_140  LW_140  GLCM1_140  Rect_140  GLCM2_140  \
0    240.15  ...       31.15    5.04       0.80      0.58       8.56   
1     57.09  ...       12.01    3.70       0.52      0.96       7.01   
2     60.48  ...       18.75    3.09       0.90      0.63       8.32   
3     61.92  ...       27.67    6.33       0.89      0.70       8.56   
4    100.64  ...       32.05    1.01       0.83      0.75       8.62   

   Dens_140  Assym_140  NDVI_140  BordLngth_140  GLCM3_140  
0      0.82       0.98     -0.10           1512    1287.52  
1      1.69       0.86     -0.14            196    2659.74  
2      1.38       0.84      0.10           1198     720.38  
3      1.10       0.96      0.20            524     891.36  
4      2.08       0.08     -0.10            496    1194.76  

[5 rows x 148 columns]

In [6]:
test_data = pd.read_csv(r'C:\Users\mhime\Downloads\test_data.csv')

In [7]:
test_data.head()

class  BrdIndx  Area  Round  Bright  Compact  ShpIndx  Mean_G  Mean_R  \
0       car      1.27    91   0.97  231.38     1.39     1.47  207.92  241.74   
1  concrete      2.36   241   1.56  216.15     2.46     2.51  187.85  229.39   
2  concrete      2.12   266   1.47  232.18     2.07     2.21  206.54  244.22   
3  concrete      2.42   399   1.28  230.40     2.49     2.73  204.60  243.27   
4  concrete      2.15   944   1.73  193.18     2.28     4.10  165.98  205.55   

   Mean_NIR  ...  SD_NIR_140  LW_140  GLCM1_140  Rect_140  GLCM2_140  \
0    244.48  ...       26.18    2.00       0.50      0.85       6.29   
1    231.20  ...       22.29    2.25       0.79      0.55       8.42   
2    245.79  ...       15.59    2.19       0.76      0.74       7.24   
3    243.32  ...       13.51    3.34       0.82      0.74       7.44   
4    208.00  ...       15.65   50.08       0.85      0.49       8.15   

   Dens_140  Assym_140  NDVI_140  BordLngth_140  GLCM3_140  
0      1.67       0.70     -0.08             56    3806.36  
1      1.38       0.81     -0.09           1746    1450.14  
2      1.68       0.81     -0.07            566    1094.04  
3      1.36       0.92     -0.09           1178    1125.38  
4      0.23       1.00     -0.08           6232    1146.38  

[5 rows x 148 columns]

b. remove any rows that have missing data across both sets of data

In [8]:
train_data.shape

(507, 148)

In [9]:
train_data= train_data.dropna(axis=0)

In [10]:
train_data.shape

(507, 148)

In [11]:
test_data.shape

(168, 148)

In [12]:
test_data= test_data.dropna(axis=0)

In [13]:
test_data.shape

(168, 148)

c. seperate out your target variable

In [14]:
y_train = train_data.iloc[:,0]

In [15]:
y_train.head()

0    concrete 
1      shadow 
2      shadow 
3        tree 
4     asphalt 
Name: class, dtype: object

In [16]:
y_test = test_data.iloc[:,0]

In [17]:
y_test.head()

0         car 
1    concrete 
2    concrete 
3    concrete 
4    concrete 
Name: class, dtype: object

In [18]:
X_train = train_data[train_data.columns[1:150]]

In [19]:
X_train.head()

BrdIndx  Area  Round  Bright  Compact  ShpIndx  Mean_G  Mean_R  Mean_NIR  \
0     1.32   131   0.81  222.74     1.66     2.18  192.94  235.11    240.15   
1     1.59   864   0.94   47.56     1.41     1.87   36.82   48.78     57.09   
2     1.41   409   1.00   51.38     1.37     1.53   41.72   51.96     60.48   
3     2.58   187   1.91   70.08     3.41     3.11   93.13   55.20     61.92   
4     2.60   116   2.05   89.57     3.06     3.02   73.17   94.89    100.64   

    SD_G  ...  SD_NIR_140  LW_140  GLCM1_140  Rect_140  GLCM2_140  Dens_140  \
0  11.24  ...       31.15    5.04       0.80      0.58       8.56      0.82   
1   8.15  ...       12.01    3.70       0.52      0.96       7.01      1.69   
2   8.11  ...       18.75    3.09       0.90      0.63       8.32      1.38   
3  28.60  ...       27.67    6.33       0.89      0.70       8.56      1.10   
4   5.41  ...       32.05    1.01       0.83      0.75       8.62      2.08   

   Assym_140  NDVI_140  BordLngth_140  GLCM3_140  
0       0.98     -0.10           1512    1287.52  
1       0.86     -0.14            196    2659.74  
2       0.84      0.10           1198     720.38  
3       0.96      0.20            524     891.36  
4       0.08     -0.10            496    1194.76  

[5 rows x 147 columns]

In [20]:
X_train.shape

(507, 147)

In [21]:
X_test = test_data[test_data.columns[1:150]]

In [22]:
X_test.head()

BrdIndx  Area  Round  Bright  Compact  ShpIndx  Mean_G  Mean_R  Mean_NIR  \
0     1.27    91   0.97  231.38     1.39     1.47  207.92  241.74    244.48   
1     2.36   241   1.56  216.15     2.46     2.51  187.85  229.39    231.20   
2     2.12   266   1.47  232.18     2.07     2.21  206.54  244.22    245.79   
3     2.42   399   1.28  230.40     2.49     2.73  204.60  243.27    243.32   
4     2.15   944   1.73  193.18     2.28     4.10  165.98  205.55    208.00   

    SD_G  ...  SD_NIR_140  LW_140  GLCM1_140  Rect_140  GLCM2_140  Dens_140  \
0  21.41  ...       26.18    2.00       0.50      0.85       6.29      1.67   
1   6.57  ...       22.29    2.25       0.79      0.55       8.42      1.38   
2   6.16  ...       15.59    2.19       0.76      0.74       7.24      1.68   
3   5.76  ...       13.51    3.34       0.82      0.74       7.44      1.36   
4  11.46  ...       15.65   50.08       0.85      0.49       8.15      0.23   

   Assym_140  NDVI_140  BordLngth_140  GLCM3_140  
0       0.70     -0.08             56    3806.36  
1       0.81     -0.09           1746    1450.14  
2       0.81     -0.07            566    1094.04  
3       0.92     -0.09           1178    1125.38  
4       1.00     -0.08           6232    1146.38  

[5 rows x 147 columns]

In [23]:
X_test.shape

(168, 147)

d. scale all the features/predictors

In [24]:
from sklearn.preprocessing import StandardScaler

In [25]:
scaler = StandardScaler()

In [26]:
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [27]:
X_train = scaler.transform(X_train)

In [28]:
X_test = scaler.transform(X_test)

In [29]:
X_train

array([[-1.14075473, -0.61823212, -0.76157526, ..., -0.95313273,
         0.10334763,  0.34780924],
       [-0.70431561,  0.43196247, -0.53002518, ..., -1.25230517,
        -1.09711859,  2.92039508],
       [-0.99527502, -0.21993185, -0.42315592, ...,  0.5427295 ,
        -0.1830858 , -0.71544325],
       ...,
       [ 0.68582378,  0.13825184,  0.25368276, ..., -0.87833962,
         1.09218151,  0.26440108],
       [-0.20321885, -0.67840698, -0.40534437, ..., -1.10271895,
        -0.61364815,  0.6910969 ],
       [ 0.73431702, -0.58384648,  0.78802909, ..., -1.02792584,
        -1.00589775,  1.66713547]])

In [30]:
X_test

array([[-1.22157678, -0.67554151, -0.47659055, ..., -0.8035465 ,
        -1.22482777,  5.07003463],
       [ 0.54034408, -0.46063129,  0.57429056, ..., -0.87833962,
         0.31680439,  0.65268303],
       [ 0.1523982 , -0.42481292,  0.41398666, ..., -0.72875339,
        -0.75960149, -0.0149197 ],
       ...,
       [ 1.52637318, -0.44487121,  0.87708681, ...,  0.24355706,
        -0.41113789, -0.52522991],
       [-0.73664444, -0.49644966,  0.05775578, ...,  0.46793639,
        -1.04421051,  0.89748944],
       [ 0.1523982 ,  0.3918459 , -0.63689445, ..., -0.87833962,
         1.47895786,  0.95528835]])

# Part 2: Random Forest Classifier - Base Model

a.  Start by creating a simple random forest using only default parameters.  Fit RandomForestClassifier to training data

In [31]:
from sklearn.ensemble import RandomForestClassifier

In [32]:
rf_clf = RandomForestClassifier()

In [33]:
rf_clf.fit(X_train, y_train)

C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

b. Use fitted model to predict on test data. Use .predict() method to get the predicted classes

In [34]:
y_pred = rf_clf.predict(X_test)

c. Calculate the confusion matrix and classification report for the test data

In [35]:
from sklearn.metrics import classification_report

In [36]:
from sklearn.metrics import confusion_matrix

In [37]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

    asphalt        0.74      1.00      0.85        14
   building        0.67      0.88      0.76        25
        car        1.00      0.87      0.93        15
   concrete        0.65      0.74      0.69        23
      grass        0.93      0.90      0.91        29
       pool        1.00      0.73      0.85        15
     shadow        1.00      0.88      0.93        16
       soil        0.83      0.36      0.50        14
       tree        0.83      0.88      0.86        17

    accuracy                           0.82       168
   macro avg       0.85      0.80      0.81       168
weighted avg       0.84      0.82      0.81       168



In [38]:
print(confusion_matrix(y_test, y_pred))

[[14  0  0  0  0  0  0  0  0]
 [ 1 22  0  2  0  0  0  0  0]
 [ 1  1 13  0  0  0  0  0  0]
 [ 0  5  0 17  0  0  0  1  0]
 [ 0  0  0  0 26  0  0  0  3]
 [ 1  3  0  0  0 11  0  0  0]
 [ 2  0  0  0  0  0 14  0  0]
 [ 0  2  0  6  1  0  0  5  0]
 [ 0  0  0  1  1  0  0  0 15]]


d. Calculate predictions for the training data and build the classification report and confusion matrix. are there signs of overfitting? why or why not?

In [39]:
y_pred_train = rf_clf.predict(X_train)

In [40]:
print(classification_report(y_train, y_pred_train))

              precision    recall  f1-score   support

    asphalt        0.98      0.98      0.98        45
   building        0.98      0.99      0.98        97
        car        1.00      1.00      1.00        21
   concrete        1.00      1.00      1.00        93
      grass        0.99      0.99      0.99        83
       pool        1.00      0.93      0.96        14
     shadow        0.98      1.00      0.99        45
       soil        1.00      1.00      1.00        20
       tree        1.00      0.99      0.99        89

    accuracy                           0.99       507
   macro avg       0.99      0.99      0.99       507
weighted avg       0.99      0.99      0.99       507



In [41]:
print(confusion_matrix(y_train, y_pred_train))

[[44  0  0  0  0  0  1  0  0]
 [ 1 96  0  0  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  0  0 93  0  0  0  0  0]
 [ 0  1  0  0 82  0  0  0  0]
 [ 0  1  0  0  0 13  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  0  0  0 20  0]
 [ 0  0  0  0  1  0  0  0 88]]


Really severe overfitting. The accuracy of my predictions for the training data is 99% and the accuracy of my predictions for the test data is 80%.  That is a huge gap

e. Identify the top 5 features. 

In [48]:
feats = {}
for feature, importance in zip(test_data.columns, rf_clf.feature_importances_):
    feats[feature] = importance

In [49]:
importances = pd.DataFrame.from_dict(feats, orient= 'index').rename(columns = {0: 'Gini-importance'})

In [51]:
importances.sort_values(by = 'Gini-importance', ascending = False)

Gini-importance
Assym              0.054467
Mean_R_40          0.048194
Assym_40           0.037491
Round              0.034209
Mean_R_100         0.031493
...                     ...
LW_140             0.000000
LW_60              0.000000
Bright             0.000000
Dens_120           0.000000
class              0.000000

[147 rows x 1 columns]

# Part 3: Linear SVM Classifier - Base Model:

a. Create a simple LinearSVC Classifier using only default parameters. Fit model on training data

In [66]:
from sklearn.svm import LinearSVC

In [67]:
svc_clf = LinearSVC()

In [68]:
svc_clf.fit(X_train, y_train)

C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\svm\base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
          verbose=0)

b. Use fitted model to predict on test data. Use .predict() method to get the predicted classes

In [69]:
y_pred_svc = svc_clf.predict(X_test)

c. Calculate the confusion matrix and classification report for the test data

In [70]:
print(classification_report(y_test, y_pred_svc))

              precision    recall  f1-score   support

    asphalt        0.76      0.93      0.84        14
   building        0.65      0.88      0.75        25
        car        0.80      0.80      0.80        15
   concrete        0.79      0.65      0.71        23
      grass        0.72      0.90      0.80        29
       pool        1.00      0.87      0.93        15
     shadow        0.93      0.88      0.90        16
       soil        1.00      0.43      0.60        14
       tree        0.77      0.59      0.67        17

    accuracy                           0.78       168
   macro avg       0.83      0.77      0.78       168
weighted avg       0.80      0.78      0.77       168



In [71]:
print(confusion_matrix(y_test, y_pred_svc))

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  1  1  1  0  0  0  0]
 [ 0  2 12  0  0  0  0  0  1]
 [ 1  6  1 15  0  0  0  0  0]
 [ 0  0  0  1 26  0  0  0  2]
 [ 1  0  1  0  0 13  0  0  0]
 [ 2  0  0  0  0  0 14  0  0]
 [ 0  4  0  1  3  0  0  6  0]
 [ 0  0  0  1  6  0  0  0 10]]


d. Calculate predictions for the training data and build the classification report and confusion matrix. are there signs of overfitting? why or why not?

In [72]:
y_pred_train_svc = svc_clf.predict(X_train)

In [73]:
print(classification_report(y_train, y_pred_train_svc))

              precision    recall  f1-score   support

    asphalt        1.00      1.00      1.00        45
   building        0.99      1.00      0.99        97
        car        1.00      1.00      1.00        21
   concrete        1.00      1.00      1.00        93
      grass        1.00      0.96      0.98        83
       pool        1.00      1.00      1.00        14
     shadow        1.00      1.00      1.00        45
       soil        1.00      1.00      1.00        20
       tree        0.98      1.00      0.99        89

    accuracy                           0.99       507
   macro avg       1.00      1.00      1.00       507
weighted avg       0.99      0.99      0.99       507



In [74]:
print(confusion_matrix(y_train, y_pred_train_svc))

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  0  0 93  0  0  0  0  0]
 [ 0  1  0  0 80  0  0  0  2]
 [ 0  0  0  0  0 14  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  0  0  0 20  0]
 [ 0  0  0  0  0  0  0  0 89]]


Really severe overfitting again. The accuracy of my predictions for the training data is 99% and the accuracy of my predictions for the test data is 78%.  An even larger gap than before

# Part 4: Support Vector Machine Classifier + Linear Kernel + Grid Search

a. Now going to use GridSearchCV to try various hyperparameters in a SVM with linear kernel.  Use SVC from sklearn with kernel = 'linear'.  Run GridSearchVC with the following  (C: 0.01-10 in increments of 0.2 (use np.arange()).  Use 5 cross-fold and the default scoring.  Set verbose = 0 to reduce the printing

In [75]:
from sklearn.model_selection import GridSearchCV

In [78]:
from sklearn.svm import SVC

In [80]:
svc = SVC(kernel = 'linear')

In [83]:
parameters = {'C': np.arange(0.01, 10, 0.2)}

In [85]:
clf = GridSearchCV(svc, parameters, cv = 5, verbose =0)

In [87]:
clf.fit(X_train, y_train)

C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\model_selection\_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='linear',
                           max_iter=-1, probability=False, random_state=None,
                           shrinking=True, tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid={'C': array([0.01, 0.21, 0.41, 0.61, 0.81, 1.01, 1.21, 1.41, 1.61, 1.81, 2.01,
       2.21, 2.41, 2.61, 2.81, 3.01, 3.21, 3.41, 3.61, 3.81, 4.01, 4.21,
       4.41, 4.61, 4.81, 5.01, 5.21, 5.41, 5.61, 5.81, 6.01, 6.21, 6.41,
       6.61, 6.81, 7.01, 7.21, 7.41, 7.61, 7.81, 8.01, 8.21, 8.41, 8.61,
       8.81, 9.01, 9.21, 9.41, 9.61, 9.81])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

b. identify the best performing model

In [89]:
clf.best_params_

{'C': 0.01}

In [90]:
svc_best = clf.best_estimator_

c. Use the best estimator model to predict on test data. Use .predict() method to get the predicted classes

In [91]:
y_pred_best = svc_best.predict(X_test)

d. Calculate the confusion matrix and classification report for the test data

In [92]:
print(classification_report(y_test, y_pred_best))

              precision    recall  f1-score   support

    asphalt        0.93      0.93      0.93        14
   building        0.71      0.88      0.79        25
        car        1.00      0.93      0.97        15
   concrete        0.65      0.74      0.69        23
      grass        0.83      0.86      0.85        29
       pool        1.00      0.93      0.97        15
     shadow        0.88      0.94      0.91        16
       soil        0.80      0.29      0.42        14
       tree        0.82      0.82      0.82        17

    accuracy                           0.82       168
   macro avg       0.85      0.81      0.82       168
weighted avg       0.83      0.82      0.81       168



In [93]:
print(confusion_matrix(y_test, y_pred_best))

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  0  2  1  0  0  0  0]
 [ 0  1 14  0  0  0  0  0  0]
 [ 0  5  0 17  0  0  0  1  0]
 [ 0  0  0  1 25  0  0  0  3]
 [ 0  0  0  0  0 14  1  0  0]
 [ 1  0  0  0  0  0 15  0  0]
 [ 0  3  0  5  2  0  0  4  0]
 [ 0  0  0  1  2  0  0  0 14]]


e. Calculate predictions for the training data and build the classification report and confusion matrix. are there signs of overfitting? why or why not?

In [94]:
y_pred_train_best = svc_best.predict(X_train)

In [95]:
print(classification_report(y_train, y_pred_train_best))

              precision    recall  f1-score   support

    asphalt        0.93      0.89      0.91        45
   building        0.85      0.90      0.87        97
        car        1.00      0.90      0.95        21
   concrete        0.87      0.89      0.88        93
      grass        0.91      0.84      0.88        83
       pool        1.00      0.86      0.92        14
     shadow        0.86      0.96      0.91        45
       soil        1.00      0.55      0.71        20
       tree        0.87      0.96      0.91        89

    accuracy                           0.89       507
   macro avg       0.92      0.86      0.88       507
weighted avg       0.89      0.89      0.89       507



In [96]:
print(confusion_matrix(y_train, y_pred_train_best))

[[40  0  0  0  0  0  5  0  0]
 [ 2 87  0  7  0  0  1  0  0]
 [ 0  1 19  1  0  0  0  0  0]
 [ 0  9  0 83  1  0  0  0  0]
 [ 0  1  0  0 70  0  0  0 12]
 [ 0  1  0  0  1 12  0  0  0]
 [ 1  0  0  0  0  0 43  0  1]
 [ 0  3  0  4  2  0  0 11  0]
 [ 0  0  0  0  3  0  1  0 85]]


Much less overfitting than previous models. The accuracy of my predictions for the training data is now 89% and the accuracy of my predictions for the test data is raised to 82%.  The gap has closed a bit 

# Part 5: Support Vector Machine Classifier + Polynomial Kernel + GridSearch

a. We will be using GridSearchCV to try various hyperparameters in a SVM with a polynomial kernel.  Use SVC from sklearn with kernel='poly'. Run GridSearchCV using the following: C: 0.01-10 in increments of 0.2, degree = 2,3,4,5,6

In [97]:
svc_poly = SVC(kernel = 'poly')

In [101]:
parameters = {'C':np.arange(0.01,10,0.2), 'degree':[2,3,4,5,6]}

In [113]:
gs_svc_poly = GridSearchCV(svc_poly, parameters, cv=5, verbose = 0)

In [114]:
gs_svc_poly.fit(X_train, y_train)

C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\svm\base.py:193: FutureWarni

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='poly', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid={'C': array([0.01, 0.21, 0.41, 0.61, 0.81, 1.01, 1.21, 1.41, 1.61, 1.81, 2.01,
       2.21, 2.41, 2.61, 2.81, 3.01, 3.21, 3.41, 3.61, 3.81, 4.01, 4.21,
       4.41, 4.61, 4.81, 5.01, 5.21, 5.41, 5.61, 5.81, 6.01, 6.21, 6.41,
       6.61, 6.81, 7.01, 7.21, 7.41, 7.61, 7.81, 8.01, 8.21, 8.41, 8.61,
       8.81, 9.01, 9.21, 9.41, 9.61, 9.81]),
                         'degree': [2, 3, 4, 5, 6]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

b. identify the best performing model

In [115]:
gs_svc_poly.best_params_

{'C': 8.41, 'degree': 3}

In [116]:
gs_svc_best = gs_svc_poly.best_estimator_

c. Use the best estimator model to predict on test data. Use .predict() method to get the predicted classes

In [117]:
y_pred_gs_svc_best = gs_svc_best.predict(X_test)

d. Calculate the confusion matrix and classification report for the test data

In [118]:
print(classification_report(y_test, y_pred_gs_svc_best))

              precision    recall  f1-score   support

    asphalt        0.87      0.93      0.90        14
   building        0.69      0.88      0.77        25
        car        1.00      0.73      0.85        15
   concrete        0.71      0.74      0.72        23
      grass        0.72      0.90      0.80        29
       pool        0.93      0.93      0.93        15
     shadow        0.88      0.88      0.88        16
       soil        0.25      0.07      0.11        14
       tree        0.87      0.76      0.81        17

    accuracy                           0.78       168
   macro avg       0.77      0.76      0.75       168
weighted avg       0.76      0.78      0.76       168



In [119]:
print(confusion_matrix(y_test, y_pred_gs_svc_best))

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  0  2  1  0  0  0  0]
 [ 0  2 11  0  0  1  0  1  0]
 [ 0  5  0 17  0  0  0  1  0]
 [ 0  0  0  0 26  0  0  1  2]
 [ 0  0  0  0  0 14  1  0  0]
 [ 2  0  0  0  0  0 14  0  0]
 [ 0  3  0  4  6  0  0  1  0]
 [ 0  0  0  1  3  0  0  0 13]]


e. Calculate predictions for the training data and build the classification report and confusion matrix. are there signs of overfitting? why or why not?

In [120]:
y_pred_train_gs_svc_best = gs_svc_best.predict(X_train)

In [121]:
print(classification_report(y_train, y_pred_train_gs_svc_best))

              precision    recall  f1-score   support

    asphalt        1.00      1.00      1.00        45
   building        0.98      1.00      0.99        97
        car        1.00      0.95      0.98        21
   concrete        1.00      0.98      0.99        93
      grass        0.90      0.99      0.94        83
       pool        1.00      0.93      0.96        14
     shadow        1.00      1.00      1.00        45
       soil        1.00      0.75      0.86        20
       tree        1.00      0.99      0.99        89

    accuracy                           0.98       507
   macro avg       0.99      0.95      0.97       507
weighted avg       0.98      0.98      0.98       507



In [122]:
print(confusion_matrix(y_train, y_pred_train_gs_svc_best))

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 20  0  1  0  0  0  0]
 [ 0  1  0 91  1  0  0  0  0]
 [ 0  1  0  0 82  0  0  0  0]
 [ 0  0  0  0  1 13  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  5  0  0 15  0]
 [ 0  0  0  0  1  0  0  0 88]]


Showing higher overfitting than the previous linear kernal SVC. The accuracy of my predictions for the training data is now 98% and the accuracy of my predictions for the test data is lowered to 78%.  The gap has widened again 

# Part 6: Support Vector Machine Classifier + RBF Kernel + GridSearch

a. We will be using GridSearchCV to try various hyperparameters in a SVM with a rbf kernel.  Use SVC from sklearn with kernel='rbf'. Run GridSearchCV using the following: C: 0.01-10 in increments of 0.2, gamma: 0.01,0.1,1,10,100

In [123]:
svc_rbf = SVC(kernel = 'rbf')

In [124]:
parameters = {'C':np.arange(0.01,10,0.2), 'gamma':[0.01,0.1,1,10,100]}

In [125]:
gs_svc_rbf = GridSearchCV(svc_rbf, parameters, cv=5, verbose = 0)

In [126]:
gs_svc_rbf.fit(X_train, y_train)

C:\Users\mhime\Anaconda2\lib\site-packages\sklearn\model_selection\_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid={'C': array([0.01, 0.21, 0.41, 0.61, 0.81, 1.01, 1.21, 1.41, 1.61, 1.81, 2.01,
       2.21, 2.41, 2.61, 2.81, 3.01, 3.21, 3.41, 3.61, 3.81, 4.01, 4.21,
       4.41, 4.61, 4.81, 5.01, 5.21, 5.41, 5.61, 5.81, 6.01, 6.21, 6.41,
       6.61, 6.81, 7.01, 7.21, 7.41, 7.61, 7.81, 8.01, 8.21, 8.41, 8.61,
       8.81, 9.01, 9.21, 9.41, 9.61, 9.81]),
                         'gamma': [0.01, 0.1, 1, 10, 100]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

b. identify the best performing model

In [127]:
gs_svc_rbf.best_params_

{'C': 4.21, 'gamma': 0.01}

In [128]:
gs_svc_rbf_best = gs_svc_rbf.best_estimator_

c. Use the best estimator model to predict on test data. Use .predict() method to get the predicted classes

In [129]:
y_pred_gs_svc_rbf = gs_svc_rbf_best.predict(X_test)

d. Calculate the confusion matrix and classification report for the test data

In [130]:
print(classification_report(y_test, y_pred_gs_svc_rbf))

              precision    recall  f1-score   support

    asphalt        0.93      0.93      0.93        14
   building        0.72      0.84      0.78        25
        car        1.00      0.87      0.93        15
   concrete        0.68      0.83      0.75        23
      grass        0.84      0.90      0.87        29
       pool        1.00      0.93      0.97        15
     shadow        0.88      0.94      0.91        16
       soil        1.00      0.36      0.53        14
       tree        0.88      0.88      0.88        17

    accuracy                           0.84       168
   macro avg       0.88      0.83      0.84       168
weighted avg       0.86      0.84      0.83       168



In [131]:
print(confusion_matrix(y_test, y_pred_gs_svc_rbf))

[[13  0  0  0  0  0  1  0  0]
 [ 0 21  0  3  1  0  0  0  0]
 [ 0  1 13  1  0  0  0  0  0]
 [ 0  4  0 19  0  0  0  0  0]
 [ 0  1  0  0 26  0  0  0  2]
 [ 0  0  0  0  0 14  1  0  0]
 [ 1  0  0  0  0  0 15  0  0]
 [ 0  2  0  4  3  0  0  5  0]
 [ 0  0  0  1  1  0  0  0 15]]


e. Calculate predictions for the training data and build the classification report and confusion matrix. are there signs of overfitting? why or why not?

In [132]:
y_pred_train_gs_svc_rbf = gs_svc_rbf_best.predict(X_train)

In [133]:
print(classification_report(y_train, y_pred_train_gs_svc_rbf))

              precision    recall  f1-score   support

    asphalt        1.00      1.00      1.00        45
   building        0.98      1.00      0.99        97
        car        1.00      1.00      1.00        21
   concrete        1.00      0.99      0.99        93
      grass        0.99      0.99      0.99        83
       pool        1.00      1.00      1.00        14
     shadow        1.00      1.00      1.00        45
       soil        1.00      1.00      1.00        20
       tree        1.00      0.99      0.99        89

    accuracy                           0.99       507
   macro avg       1.00      1.00      1.00       507
weighted avg       0.99      0.99      0.99       507



In [134]:
print(confusion_matrix(y_train, y_pred_train_gs_svc_rbf))

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  1  0 92  0  0  0  0  0]
 [ 0  1  0  0 82  0  0  0  0]
 [ 0  0  0  0  0 14  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  0  0  0 20  0]
 [ 0  0  0  0  1  0  0  0 88]]


There are signs of overfitting but less than the previous polynomial kernal SVC. The accuracy of my predictions for the training data is now 99% and the accuracy of my predictions for the test data is raised to 84%.  The gap has closed a bit but still significant signs of overfitting

# Part 7: Conceptual Questions

a) From the models run in steps 2-6, which performs the best based on the Classification Report? Support your reasoning with evidence around your test data. 


My SVMClassifier + linear kernel + gridserach model from Part 4 performed best based on the classification report.  It had a relatively high overall accuracy on the train data (82%) and generalized better than all the other models with a 89% accuracy on the test data.  It also had relatively high weighted avg precision (83%) and recall(82%) scores on the training data.  All models had trouble generalizing most likely due to a relatively small training set. I would recommend obtaining more training data to improve the performance of these models

b) Compare models run for steps 4-6 where different kernels were used. What is the benefit of using a polynomial or rbf kernel over a linear kernel? What could be a downside of using a polynomial or rbf kernel? 


Polynomial can be easy to implement but a low polynomial degreee can't deal with very complex datasets and a high polynomial degree creates a huge number of features making the model too slow.  RBF may also be computationally expensive to compute as well.  But they can be more effective than linear kernel when classes are easily seperable.

c) Explain the 'C' parameter used in steps 4-6. What does a small C mean versus a large C in sklearn? Why is it important to use the 'C' parameter when fitting a model? 


Parameter C is the inverse of the strength of regularization (penalty).   It also controls the trade off between a smooth decision boundary and classifying the training points correctly. So a high value of C will tell the model to choose a smaller-margin and does a better job of getting all the training points classified correctly.  A snaller C value would cause the model to create a larger seperating margin and allow more missclassification.  As the value of c increases, the model overfits, as it descreases, the model underfits. So its important to complete a gridsearch to find the optimal value for this parameter.

d) Scaling our input data does not matter much for Random Forest, but it is a critical step for Support Vector Machines. Explain why this is such a critical step. Also, provide an example of a feature from this data set that could cause issues with our SVMs if not scaled.


SVM are concerned with the distance between data pionts.  When using kernels in SVM, you are performing some sort of transformation to make the data higher dimension than it previously was.  In polynomial for example, you are completing a dot product between features.  If feature values are not scaled, you could be multiplying a very large value such as training data column GLCM3_140 row #1 : 1287 by a very small value such as column Compact row #1 1.66. That produces a number that does not conceptually make sense and makes it difficult to effectively measure distance between data points and make classifications and decision boundaries.

It would not make sense to compare the Round variable (shape variable) to Mean_G (Spectral variable, amount of green in an image).

e) Describe conceptually what the purpose of a kernel is for Support Vector Machines.


When datasets are not linearly seperable, kernels become necessary to make effective decision boundaries between classes.  One approach to fixing this would be to add more features such as polynomial features. But with a high number of features or complex datasets, this becomes very computationally expensive. A kernel is capable of transforming the dataset into a higher dimensional space and providing a similar result without the computation expense.